### DISCUSSION TITLE
[Question] How does the topological `pnpm run` strategy work

### DISCUSSION THREAD

**Discussion by michaelwarren1106** [pnpm/pnpm#5204] [General]
I have a monorepo that I’m testing switching to pnpm, and my packages have several cross-pkg dependencies. 

For example, `package/a` depends on `core` and package/a won’t build correctly unless core has been built first. 

I’ve tried to run `pnpm -r build` with the understanding that pnpm will look at the dependencies of a package and run the same script on dependencies first. In my example, this would mean that core’s build script gets run first then the `build` script for `package/a`.

But i don’t see this happening, and the topological order of script runs isn’t very well documented. 

Another related question is about symlinking those same dependencies. when i run `pnpm i` at the root and i’m using `linkDirectory:true` to symlinking each of my packages `dist` folders, it seems that pnpm won’t create the symlink unless the folder exists. Is there an option to create the symlink anyway since the dist folder will exist later?

**Comment by zkochan** [comment]
Yes, it should run the scripts in topological order, unless you use the [--no-sort](https://pnpm.io/cli/recursive#--no-sort) option. Are your local deps added using the [workspace protocol](https://pnpm.io/workspaces#workspace-protocol-workspace)?

> Another related question is about symlinking those same dependencies. when i run pnpm i at the root and i’m using linkDirectory:true to symlinking each of my packages dist folders, it seems that pnpm won’t create the symlink unless the folder exists. Is there an option to create the symlink anyway since the dist folder will exist later?

I guess it should work. If it doesn't work, it is probably a bug.

**Comment by michaelwarren1106** [comment]
not yet. my dependencies are just using package name and version as normal, not `workspace:<version>`

**Comment by michaelwarren1106** [comment]
is the workspace protocol required for topological ordering?

**Reply by zkochan** [reply]
I don't think it is required. But without the workspace protocol the package might be installed from the registry, if there is a newer version in the registry, matching the range. So it is better to use the workspace protocol.

**Reply by michaelwarren1106** [reply]
i’ve got concerns about using the workspace protocol in my workflow. i’m using the `publishConfig.directory` and `publishConfig.linkDirectory` features to only deploy the contents of my built `dist` folder for all my packages. 

when i symlink across the monorepo i need to link those `dist` folders locally and i need to copy in my `package.json` file from the package root at publish time. if i use the workspace protocol, then i don’t think the timing will be right with the version/publish steps? i’m also using changesets to version and publish (not the github action, a similar but custom flow on azure devops build pipelines) does pnpm adjust the dependency versions during the `pnpm changeset version` command? or during `pnpm changeset publish`? 

if the version numbers are changed during version, i could copy over the updated package.json to the dist folder before the publish command, but i think i saw that it’s only on the publish command. 

so that leads me to believe that i can’t really use the workspace protocol. as long as topological build scripts work it shouldn’t be a problem install-wise because there won’t be a package version that gets installed from the npm registry instead of local.

**Reply by zkochan** [reply]
The package.json in the dist folder may contain the workspace protocol entries. pnpm will replace them when packing the package.

**Reply by michaelwarren1106** [reply]
ok cool! so i just need to copy it in there before the changeset commands?

what about symlinking before the dist folders are created? is there a way to tell pnpm to create the symlink to dist even if the folder doesn’t exist so that when builds are run later the symlinks are already there?

i think the symlinking to non-existent linked directories is my last hurdle

**Reply by zkochan** [reply]
> so i just need to copy it in there before the changeset commands?

not before. After

> is there a way to tell pnpm to create the symlink to dist even if the folder doesn’t exist so that when builds are run later the symlinks are already there?

it should already work like that. If it doesn't, then it is a bug

**Reply by michaelwarren1106** [reply]
@zkochan I just tested in a clean repo (no `node_modules` or `dist` folders)

All my packages have:

```
publishConfig: {
  directory: 'dist',
  linkDirectory: true
}
```

when i run `pnpm i`, i get warnings like:

``` 
WARN  Local dependency not found at <path>/base/dist
 WARN  Local dependency not found at <path>/tokens/dist
 WARN  Local dependency not found at<path>/style/dist
```

i then manually went to my `base` package and built it, and checked another package that depends on base, and the symlink was NOT created and there was no linked package in `node_modules`

so I think we've got a bug. Is this enough info to test with, or do you want me to put together a full repro?

**Reply by michaelwarren1106** [reply]
i tested with and without the workspace protocol and neither worked

**Comment by mkstix6** [comment]
Could we have a little documentation on how the "topological" algorithm works?
And if there is any setup or prerequisites?

It would be nice to read something that confirms I understand it correctly.
More specifically I'd like to feel confident that topological sorting does what I think for a command such as…
`pnpm -F packageA -F packageB -F packageC build`
…where:

1. the order packages `build` matters
2. dependencies listed in **package.json**
3. `build` scripts for each package listed in **package.json** scripts

Using the documentation's search for "topological" I do see references to it in terms of _"use this flag to turn it off if you want"_ which is a bit of an understanding-by-deduction exercise.

Thanks very much ✨. Really great work on `pnpm`.

**Reply by shanefontaine** [reply]
Looking at the code, I do not believe the lifecycle scripts respect by the topological sorting algorithm. I'll walk through the relevant code (as I understand it) below. @zkochan I'd greatly appreciate it if you could confirm the accuracy of these observations or correct any misinterpretations. 

1. Installing workspace packages with `pnpm install` runs [`installDeps()`](https://github.com/pnpm/pnpm/blob/e363af8e1305c69b288f44aab3ae21ad9b4dc581/pkg-manager/plugin-commands-installation/src/installDeps.ts#L117).
2. This function [ignores all lifecycle scripts](https://github.com/pnpm/pnpm/blob/e363af8e1305c69b288f44aab3ae21ad9b4dc581/pkg-manager/plugin-commands-installation/src/installDeps.ts#L227-L229) while [installing deps](https://github.com/pnpm/pnpm/blob/e363af8e1305c69b288f44aab3ae21ad9b4dc581/pkg-manager/plugin-commands-installation/src/installDeps.ts#L322). It then calls [`rebuildProjects()`](https://github.com/pnpm/pnpm/blob/e363af8e1305c69b288f44aab3ae21ad9b4dc581/pkg-manager/plugin-commands-installation/src/installDeps.ts#L326) with the intention of now installing the lifecycle scripts.
3. This function calls [`runLifecycleHooksConcurrently()`](https://github.com/pnpm/pnpm/blob/main/exec/plugin-commands-rebuild/src/implementation/index.ts#L175) which appears to run the lifecycle scripts for each package in parallel (but does run each individual package's scripts [serially](https://github.com/pnpm/pnpm/blob/main/exec/lifecycle/src/runLifecycleHooksConcurrently.ts#L67-L71), as you would expect).

Happy to open a PR for code or docs about this to help future readers, but want to confirm that my understanding aligns with the intended behavior of `pnpm`. Thank you for your work 🙏

I recognize this is an [unsolvable problem](https://github.com/npm/cli/issues/3034#issuecomment-885290017) but hope to get some clarity.
